// @clawdactual/openclaw-recall/index.ts
/**
 * openclaw-recall — Automatic associative memory recall for OpenClaw agents.
 *
 * Surfaces relevant context from memory before the model processes each message.
 * Works alongside any context engine (LCM, legacy, etc.) without replacing it.
 *
 * Integration: Uses the typed `before_prompt_build` plugin hook — the explicit
 * capability registration path for prompt injection. Returns { prependContext }
 * which OpenClaw natively injects into the model's context window.
 */
import { mkdirSync, existsSync, appendFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { homedir } from "node:os";
import type { OpenClawPluginApi } from "openclaw/plugin-sdk";
import { resolveRecallConfig } from "./src/config.js";
import { extractQueries } from "./src/query-extractor.js";
import { RecallEngine } from "./src/recall-engine.js";
import { MemorySearchSource } from "./src/sources/memory-search.js";
import { LcmSearchSource } from "./src/sources/lcm-search.js";
import { QmdSearchSource } from "./src/sources/qmd-search.js";
import { EngramSearchSource } from "./src/sources/engram-search.js";

const recallPlugin = {
  id: "openclaw-recall",
  name: "Recall",
  description:
    "Automatic associative memory recall — surfaces relevant context before each message",

  configSchema: {
    parse(value: unknown) {
      const raw =
        value && typeof value === "object" && !Array.isArray(value)
          ? (value as Record<string, unknown>)
          : {};
      return resolveRecallConfig(raw);
    },
  },

  register(api: OpenClawPluginApi) {
    const pluginConfig =
      api.pluginConfig && typeof api.pluginConfig === "object" && !Array.isArray(api.pluginConfig)
        ? (api.pluginConfig as Record<string, unknown>)
        : undefined;

    const config = resolveRecallConfig(pluginConfig);

    if (!config.enabled) {
      api.logger.info("[recall] Plugin disabled via config");
      return;
    }

    // If memory-lancedb is active, skip only the memory-search source
    // (lancedb handles its own recall for that layer). Other sources
    // (LCM, QMD) still run since lancedb doesn't cover them.
    const lancedbActive = config.skipMemorySearchIfLancedb && isLancedbActive(api.config);
    if (lancedbActive) {
      config.sources.memorySearch.enabled = false;
      api.logger.info("[recall] memory-lancedb is active — disabling memory-search source (other sources still active)");
    }

    const engine = new RecallEngine(config, {
      info: (msg) => api.logger.info(msg),
      debug: (msg) => api.logger.debug?.(msg) ?? api.logger.info(msg),
      warn: (msg) => api.logger.warn(msg),
      error: (msg) => api.logger.error(msg),
    });

    // Phase 1: Register memory_search source
    // Uses OpenClaw's CLI (`openclaw memory search --json`) via runCommandWithTimeout.
    // This is the most portable approach — works with any OpenClaw install, any
    // memory backend (builtin, QMD), and doesn't depend on internal APIs.
    if (config.sources.memorySearch.enabled) {
      const memoryApi = createCliMemorySearchApi(api, config.sources.memorySearch.maxResults);
      engine.addSource(new MemorySearchSource(memoryApi, config.sources.memorySearch.maxResults));
      api.logger.info("[recall] memory-search source registered (CLI backend)");
    }

    // Phase 2: Register LCM source (FTS5 search on conversation history)
    if (config.sources.lcm.enabled) {
      const lcmDbPath = resolveLcmDbPath(api);
      const lcmSource = new LcmSearchSource(
        createCliSqliteApi(api, lcmDbPath),
        config.sources.lcm,
      );
      engine.addSource(lcmSource);
      api.logger.info(
        `[recall] LCM source registered (scope=${config.sources.lcm.scope}, db=${lcmDbPath})`,
      );
    }

    // Phase 2: Register QMD source (BM25 search via qmd CLI)
    if (config.sources.qmd.enabled) {
      const qmdSource = new QmdSearchSource(
        createCliQmdApi(api, config.sources.qmd.command),
        config.sources.qmd.maxResults,
      );
      engine.addSource(qmdSource);
      api.logger.info(`[recall] QMD source registered (command=${config.sources.qmd.command})`);
    }

    // Engram (knowledge graph) — runs as a SEPARATE parallel search,
    // not mixed into the RecallEngine source pool. Graph context is structural
    // (nodes + edges) and shouldn't compete on score with text-based recall.
    // Gets its own section: "## 🔗 Knowledge Graph"
    const engramEnabled = config.sources.engram.enabled ||
      (config.sources.engram.enabled !== false && autoDetectEngram());
    let engramSource: EngramSearchSource | null = null;
    if (engramEnabled) {
      engramSource = new EngramSearchSource(
        createCliEngramApi(api, config.sources.engram.command, config.sources.engram.excludeModels),
        config.sources.engram.maxResults,
        config.sources.engram.excludeModels,
      );
      api.logger.info(
        `[recall] engram source registered (separate section, command=${config.sources.engram.command}, max=${config.sources.engram.maxResults})`,
      );
    }

    // Use the typed `before_prompt_build` hook — the explicit capability
    // registration path for prompt injection plugins. Returns { prependContext }
    // which OpenClaw natively injects into the model's context.
    api.on("before_prompt_build", async (event, ctx) => {
      const prompt = event.prompt;
      if (!prompt || prompt.length < 3) return;

      try {
        // Extract queries ONCE, then use them for both text recall and engram.
        // RecallEngine.recall() does its own extraction internally for text sources.
        // For engram, we need to pass extracted queries since it no longer does
        // its own extraction (v0.3: removed double-extraction).
        const { queries: extractedQueries } = extractQueries(prompt);

        // Run text recall and engram search IN PARALLEL
        // Engram gets searched once per extracted query, results merged
        const [recallResult, engramSnippets] = await Promise.all([
          engine.recall(prompt),
          engramSource && extractedQueries.length > 0
            ? (async () => {
                const jobs = extractedQueries.slice(0, 3).map(q =>
                  engramSource!.search(q, config.minRelevanceScore).catch((err) => {
                    api.logger.warn(`[recall] Engram search failed: ${err instanceof Error ? err.message : err}`);
                    return [];
                  })
                );
                const allResults = (await Promise.all(jobs)).flat();

                // Deduplicate by reference, sort by score, take top N.
                // With relevance scoring (v0.4), scores determine ranking —
                // no need for round-robin interleaving.
                const seen = new Set<string>();
                const deduped = allResults.filter(s => {
                  if (seen.has(s.reference)) return false;
                  seen.add(s.reference);
                  return true;
                });
                return deduped
                  .sort((a, b) => b.score - a.score)
                  .slice(0, config.sources.engram.maxResults);
              })()
            : Promise.resolve([]),
        ]);

        // Build the combined block
        const blocks: string[] = [];

        // 1. Text recall section
        if (recallResult && recallResult.snippetCount > 0) {
          blocks.push(recallResult.block);
        }

        // 2. Knowledge graph section (separate, not competing)
        if (engramSnippets.length > 0) {
          const engramLines = [
            "",
            "## 🔗 Knowledge Graph",
            "",
          ];
          for (const snippet of engramSnippets) {
            engramLines.push(`- **[${snippet.reference}]** ${snippet.text}`);
          }
          blocks.push(engramLines.join("\n"));
        }

        if (blocks.length === 0) return;

        const fullBlock = blocks.join("\n");
        const totalSnippets = (recallResult?.snippetCount ?? 0) + engramSnippets.length;
        const estimatedTokens = Math.ceil(fullBlock.length / 4);

        if (config.debug.enabled) {
          writeDebugLog(api, config.debug.logPath, {
            timestamp: new Date().toISOString(),
            sessionKey: ctx.sessionKey,
            trigger: ctx.trigger,
            prompt: prompt.slice(0, 200),
            queryStrategy: recallResult?.queryStrategy ?? "none",
            snippetCount: totalSnippets,
            recallSnippets: recallResult?.snippetCount ?? 0,
            engramSnippets: engramSnippets.length,
            totalResults: recallResult?.totalResults ?? 0,
            estimatedTokens,
            durationMs: recallResult?.durationMs ?? 0,
          });
        }

        api.logger.info(
          `[recall] Recalled ${recallResult?.snippetCount ?? 0} text + ${engramSnippets.length} engram snippets (${estimatedTokens} est. tokens)`,
        );

        return { prependContext: fullBlock };
      } catch (err) {
        api.logger.error(
          `[recall] Error during recall: ${err instanceof Error ? err.message : err}`,
        );
        // Never break the agent flow on recall failure
        return;
      }
    });

    // Register /recall command for status and diagnostics.
    // Also satisfies explicit capability registration (no longer "hook-only").
    api.registerCommand({
      name: "recall",
      description: "Show recall plugin status and configuration",
      acceptsArgs: true,
      requireAuth: true,
      handler: (ctx) => {
        const sources = [];
        if (config.sources.memorySearch.enabled) sources.push(`memory-search (max ${config.sources.memorySearch.maxResults})`);
        if (config.sources.lcm.enabled) sources.push(`lcm (scope=${config.sources.lcm.scope}, max ${config.sources.lcm.maxResults})`);
        if (config.sources.qmd.enabled) sources.push(`qmd (max ${config.sources.qmd.maxResults})`);
        if (engramEnabled) sources.push(`engram (separate section, max ${config.sources.engram.maxResults})`);

        const lines = [
          `**Recall Plugin** v0.2.0`,
          `Status: ${config.enabled ? "✅ active" : "❌ disabled"}`,
          `Max snippets: ${config.maxSnippets}`,
          `Min relevance: ${config.minRelevanceScore}`,
          `Sources: ${sources.length > 0 ? sources.join(", ") : "none"}`,
          `Skip memory-search if lancedb: ${config.skipMemorySearchIfLancedb ? "yes" : "no"}`,
          `Exclude files: ${config.excludeFiles.length}`,
          `Exclude patterns: ${config.excludePatterns.length}`,
          `Debug: ${config.debug.enabled ? "on" : "off"}`,
        ];

        return { text: lines.join("\n") };
      },
    });

    api.logger.info(
      `[recall] Plugin loaded (maxSnippets=${config.maxSnippets}, text sources=${countEnabledSources(config) - (engramEnabled ? 1 : 0)}, engram=${engramEnabled ? "separate section" : "off"})`,
    );
  },
};

/**
 * Create a memory search API backed by OpenClaw's CLI.
 *
 * Calls `openclaw memory search <query> --json --max-results N`
 * via api.runtime.system.runCommandWithTimeout. This is portable across
 * any OpenClaw install and memory backend (builtin Voyage, QMD, etc.).
 */
function createCliMemorySearchApi(
  api: OpenClawPluginApi,
  defaultMaxResults: number,
) {
  return {
    search: async (
      query: string,
      maxResults?: number,
      minScore?: number,
    ) => {
      try {
        const limit = maxResults ?? defaultMaxResults;
        const argv = [
          "openclaw",
          "memory", "search",
          "--json",
          "--max-results", String(limit),
          ...(minScore !== undefined ? ["--min-score", String(minScore)] : []),
          "--query", query,
        ];

        const result = await api.runtime.system.runCommandWithTimeout(
          argv,
          { timeoutMs: 5000 },
        );

        if (!result.stdout) return [];

        const parsed = JSON.parse(result.stdout) as {
          results?: Array<{
            path: string;
            snippet: string;
            score: number;
            startLine?: number;
            endLine?: number;
            source?: string;
          }>;
        };

        return (parsed.results ?? []).map((r) => ({
          path: r.path,
          text: r.snippet,
          score: r.score,
          lineStart: r.startLine,
          lineEnd: r.endLine,
        }));
      } catch (err) {
        api.logger.warn(
          `[recall] CLI memory search failed: ${err instanceof Error ? err.message : err}`,
        );
        return [];
      }
    },
  };
}

/**
 * Write debug log entry (append mode).
 */
function writeDebugLog(
  api: OpenClawPluginApi,
  logPath: string,
  entry: Record<string, unknown>,
): void {
  try {
    const fullPath = api.resolvePath(logPath);
    const dir = dirname(fullPath);
    if (!existsSync(dir)) {
      mkdirSync(dir, { recursive: true });
    }
    appendFileSync(fullPath, JSON.stringify(entry) + "\n", "utf8");
  } catch {
    // Debug logging should never break anything
  }
}

function countEnabledSources(config: ReturnType<typeof resolveRecallConfig>): number {
  let count = 0;
  if (config.sources.memorySearch.enabled) count++;
  if (config.sources.qmd.enabled) count++;
  if (config.sources.lcm.enabled) count++;
  if (config.sources.engram.enabled) count++;
  return count;
}

/**
 * Detect if memory-lancedb plugin is active.
 * If it is, there's no need for recall — lancedb handles its own auto-recall.
 */
function isLancedbActive(config: unknown): boolean {
  const entries = (config as any)?.plugins?.entries;
  if (!entries) return false;
  const lancedb = entries["memory-lancedb"];
  return lancedb && lancedb.enabled !== false;
}

/**
 * Resolve the LCM database path from config or use default.
 */
function resolveLcmDbPath(api: OpenClawPluginApi): string {
  const lcmConfig = (api.config as any)?.plugins?.entries?.["lossless-claw"];
  if (lcmConfig?.db && typeof lcmConfig.db === "string") {
    return lcmConfig.db;
  }
  return api.resolvePath("~/.openclaw/lcm.db");
}

/**
 * Create a SQLite query API backed by the sqlite3 CLI.
 */
function createCliSqliteApi(api: OpenClawPluginApi, dbPath: string) {
  return {
    query: async (
      sql: string,
      timeoutMs: number = 5000,
    ): Promise<Record<string, unknown>[]> => {
      try {
        const result = await api.runtime.system.runCommandWithTimeout(
          ["sqlite3", "-json", dbPath, sql],
          { timeoutMs },
        );
        if (!result.stdout?.trim()) {
          if (result.stderr?.trim()) {
            api.logger.warn(`[recall] sqlite3 stderr: ${result.stderr.trim().slice(0, 200)}`);
          }
          return [];
        }
        return JSON.parse(result.stdout);
      } catch (err) {
        api.logger.warn(
          `[recall] sqlite3 query failed: ${err instanceof Error ? err.message : err}`,
        );
        return [];
      }
    },
  };
}

/**
 * Create a QMD search API backed by the qmd CLI.
 */
function createCliQmdApi(api: OpenClawPluginApi, command: string) {
  return {
    search: async (query: string): Promise<unknown[]> => {
      try {
        const cmdParts = command.split(/\s+/);
        const argv = [...cmdParts, query, "--json"];
        const result = await api.runtime.system.runCommandWithTimeout(argv, {
          timeoutMs: 10000,
        });
        if (!result.stdout?.trim()) return [];
        return JSON.parse(result.stdout);
      } catch (err) {
        api.logger.warn(
          `[recall] QMD search failed: ${err instanceof Error ? err.message : err}`,
        );
        return [];
      }
    },
  };
}

/**
 * Create an engram search API backed by the engram CLI.
 *
 * Calls `engram search <query> --json --limit N [--exclude ...]`
 * via api.runtime.system.runCommandWithTimeout.
 */
function createCliEngramApi(
  api: OpenClawPluginApi,
  command: string,
  defaultExcludes: string[],
) {
  return {
    search: async (query: string, excludeModels?: string[]) => {
      try {
        const excludes = excludeModels ?? defaultExcludes;
        const excludeArgs = excludes.flatMap((m) => ["--exclude", m]).join(" ");
        // Command may contain shell operators (&&, cd) so run through sh -c
        const shellCmd = `${command} search ${JSON.stringify(query)} --json --limit 5 ${excludeArgs}`;
        const argv = ["sh", "-c", shellCmd];

        const result = await api.runtime.system.runCommandWithTimeout(argv, {
          timeoutMs: 8000,
        });

        if (!result.stdout?.trim()) return [];

        const parsed = JSON.parse(result.stdout);
        return parsed.results ?? parsed ?? [];
      } catch (err) {
        api.logger.warn(
          `[recall] engram search failed: ${err instanceof Error ? err.message : err}`,
        );
        return [];
      }
    },
  };
}

/**
 * Auto-detect engram availability by checking for the database file.
 */
function autoDetectEngram(): boolean {
  const dbPath = join(homedir(), ".config", "engram", "models.db");
  return existsSync(dbPath);
}

export default recallPlugin;
