// @clawdactual/openclaw-recall/src/sources/engram-search.ts
/**
 * Engram Search source — searches knowledge graphs via the `mm` CLI.
 *
 * Surfaces structural knowledge (nodes, edges, types) from Engram graphs
 * to complement text-based recall sources like memory-search and QMD.
 *
 * v0.3: Removed internal extractQueries call — the RecallEngine already
 * extracts queries and calls search() once per query. Double-extracting
 * was degrading results (extracting queries from already-extracted phrases).
 * Also improved: CamelCase node names now match split variants via the
 * improved extractProperNouns in query-extractor.
 */

import type { RecallSnippet, RecallSource } from "../types.js";

export interface EngramApi {
  search(query: string, excludeModels?: string[]): Promise<EngramSearchResult[]>;
}

export interface EngramSearchResult {
  model: string;
  modelType: string;
  node: { label: string; type: string | null; metadata: Record<string, unknown> };
  edges: Array<{
    direction: "in" | "out";
    relationship: string;
    source?: string;
    target?: string;
  }>;
}

export class EngramSearchSource implements RecallSource {
  readonly name = "engram";

  constructor(
    private api: EngramApi,
    private maxResults: number = 3,
    private excludeModels: string[] = [],
  ) {}

  async search(query: string, minScore: number): Promise<RecallSnippet[]> {
    try {
      // The RecallEngine already extracts queries and calls us once per query.
      // We search directly with the provided query — no double-extraction.
      if (!query || query.trim().length === 0) return [];

      const results = await this.api.search(query, this.excludeModels)
        .catch(() => [] as EngramSearchResult[]);

      if (results.length === 0) return [];

      // Deduplicate by model+node label (same query can match multiple edges of same node)
      const seen = new Set<string>();
      const deduped: EngramSearchResult[] = [];
      for (const r of results) {
        const key = `${r.model}/${r.node.label}`;
        if (seen.has(key)) continue;
        seen.add(key);
        deduped.push(r);
      }

      // Score each result by query-term overlap with node label, metadata, and edge labels.
      // This replaces the fixed 0.5 score, filtering out irrelevant results.
      const queryTerms = extractQueryTerms(query);
      const scored = deduped.map((r) => ({
        result: r,
        score: scoreEngramResult(r, queryTerms),
      }));

      // Filter below minScore, sort by score descending
      return scored
        .filter((s) => s.score >= minScore)
        .sort((a, b) => b.score - a.score)
        .slice(0, this.maxResults)
        .map((s) => ({
          source: this.name,
          path: `engram/${s.result.model}`,
          text: formatSnippetText(s.result),
          score: s.score,
          reference: `engram:${s.result.model}/${s.result.node.label}`,
        }));
    } catch {
      // Source failure should not break recall — return empty
      return [];
    }
  }
}

/**
 * Extract meaningful terms from a query for relevance scoring.
 * Lowercases, strips stop words, splits CamelCase, handles hyphens.
 */
function extractQueryTerms(query: string): Set<string> {
  const trivialWords = new Set([
    "a", "an", "the", "and", "or", "but", "in", "on", "at", "to", "for",
    "of", "with", "by", "from", "is", "it", "was", "be", "are", "been",
    "have", "has", "had", "do", "does", "did", "will", "would", "could",
    "should", "may", "might", "can", "not", "no", "this", "that", "what",
    "how", "who", "which", "about", "tell", "me", "show", "get",
  ]);

  const terms = new Set<string>();
  // Split on whitespace and punctuation, normalize
  const raw = query.toLowerCase().replace(/[^a-z0-9\s-]/g, " ").split(/\s+/);
  for (const word of raw) {
    if (word.length < 2 || trivialWords.has(word)) continue;
    terms.add(word);
    // Also add hyphen-split parts: "fleet-rest" → "fleet", "rest"
    if (word.includes("-")) {
      for (const part of word.split("-")) {
        if (part.length >= 2 && !trivialWords.has(part)) terms.add(part);
      }
    }
  }

  // Also split CamelCase from the original query (before lowercasing)
  const camelParts = query.replace(/([a-z])([A-Z])/g, "$1 $2").toLowerCase().split(/\s+/);
  for (const part of camelParts) {
    if (part.length >= 2 && !trivialWords.has(part)) terms.add(part);
  }

  return terms;
}

/**
 * Score an engram result by how well it matches the query terms.
 *
 * Checks query term overlap against:
 * - Node label (highest weight: 0.4 per match)
 * - Edge labels — targets and sources (medium weight: 0.15 per match)
 * - Model name (low weight: 0.1 per match)
 * - Metadata values (low weight: 0.1 per match)
 *
 * Returns 0.0–1.0. A result with zero term overlap gets 0.0.
 */
function scoreEngramResult(result: EngramSearchResult, queryTerms: Set<string>): number {
  if (queryTerms.size === 0) return 0;

  let score = 0;
  let matchedTerms = 0;

  // Tokenize the node label (split CamelCase and hyphens)
  const labelTokens = tokenize(result.node.label);
  // Tokenize model name
  const modelTokens = tokenize(result.model);
  // Tokenize edge targets and sources
  const edgeTokens = new Set<string>();
  for (const edge of result.edges) {
    if (edge.target) tokenize(edge.target).forEach(t => edgeTokens.add(t));
    if (edge.source) tokenize(edge.source).forEach(t => edgeTokens.add(t));
    // Relationship name itself
    tokenize(edge.relationship).forEach(t => edgeTokens.add(t));
  }
  // Tokenize metadata values
  const metaTokens = new Set<string>();
  for (const val of Object.values(result.node.metadata)) {
    if (typeof val === "string") {
      tokenize(val).forEach(t => metaTokens.add(t));
    }
  }

  for (const term of queryTerms) {
    // Label match (highest signal)
    if (labelTokens.has(term)) {
      score += 0.4;
      matchedTerms++;
      continue; // Don't double-count
    }
    // Edge label match (medium signal)
    if (edgeTokens.has(term)) {
      score += 0.15;
      matchedTerms++;
      continue;
    }
    // Model name match (low signal)
    if (modelTokens.has(term)) {
      score += 0.1;
      matchedTerms++;
      continue;
    }
    // Metadata match (low signal)
    if (metaTokens.has(term)) {
      score += 0.1;
      matchedTerms++;
      continue;
    }
  }

  // Normalize: cap at 1.0, and apply a coverage factor
  // (matching 1 of 5 terms is worth less than matching 3 of 3)
  const coverage = matchedTerms / queryTerms.size;
  const rawScore = Math.min(score, 1.0);

  // Blend raw score with coverage: high overlap + high signal = best
  return Math.min(1.0, rawScore * 0.7 + coverage * 0.3);
}

/**
 * Tokenize a label/string into lowercase terms.
 * Splits CamelCase, hyphens, underscores.
 */
function tokenize(text: string): Set<string> {
  const tokens = new Set<string>();
  // Split CamelCase
  const split = text.replace(/([a-z])([A-Z])/g, "$1 $2");
  // Split on non-alphanumeric
  for (const part of split.toLowerCase().split(/[^a-z0-9]+/)) {
    if (part.length >= 2) tokens.add(part);
  }
  return tokens;
}

/**
 * Format a search result into a human-readable snippet.
 *
 * Example output:
 *   Model: chitin (code) | RetrievalEngine (module)
 *     → uses: EmbeddingStore, context-detect, decay
 *     ← used_by: index.ts
 */
function formatSnippetText(result: EngramSearchResult): string {
  const typeStr = result.node.type ? ` (${result.node.type})` : "";
  const modelStr = result.modelType ? ` (${result.modelType})` : "";
  const parts: string[] = [
    `Model: ${result.model}${modelStr} | ${result.node.label}${typeStr}`,
  ];

  const outgoing = result.edges.filter((e) => e.direction === "out");
  const incoming = result.edges.filter((e) => e.direction === "in");

  if (outgoing.length > 0) {
    const grouped = groupEdges(outgoing, "target");
    for (const [rel, targets] of grouped) {
      parts.push(`→ ${rel}: ${targets.join(", ")}`);
    }
  }

  if (incoming.length > 0) {
    const grouped = groupEdges(incoming, "source");
    for (const [rel, sources] of grouped) {
      parts.push(`← ${rel}: ${sources.join(", ")}`);
    }
  }

  return parts.join(" ");
}

/**
 * Group edges by relationship and collect their target/source labels.
 */
function groupEdges(
  edges: EngramSearchResult["edges"],
  labelKey: "source" | "target",
): Map<string, string[]> {
  const grouped = new Map<string, string[]>();
  for (const edge of edges) {
    const label = edge[labelKey] ?? "?";
    const rel = edge.relationship;
    if (!grouped.has(rel)) grouped.set(rel, []);
    grouped.get(rel)!.push(label);
  }
  return grouped;
}
