// @clawdactual/openclaw-recall/src/sources/lcm-search.ts
/**
 * LCM Search source — searches OpenClaw's LCM conversation history via SQLite FTS5.
 *
 * Uses the sqlite3 CLI (via runCommandWithTimeout) to query FTS5 tables
 * in the LCM database. Supports searching messages, summaries, or both.
 */

import type { RecallSnippet, RecallSource } from "../types.js";

export interface SqliteApi {
  query(sql: string, timeoutMs?: number): Promise<Record<string, unknown>[]>;
}

export interface LcmSearchConfig {
  maxResults: number;
  scope: "messages" | "summaries" | "both";
}

export class LcmSearchSource implements RecallSource {
  readonly name = "lcm";
  private db: SqliteApi;
  private config: LcmSearchConfig;

  constructor(db: SqliteApi, config: LcmSearchConfig) {
    this.db = db;
    this.config = config;
  }

  async search(query: string, minScore: number): Promise<RecallSnippet[]> {
    const ftsQuery = sanitizeFtsQuery(query);
    if (!ftsQuery) return [];

    const results: RecallSnippet[] = [];

    try {
      if (this.config.scope === "messages" || this.config.scope === "both") {
        const messages = await this.searchMessages(ftsQuery);
        results.push(...messages);
      }

      if (this.config.scope === "summaries" || this.config.scope === "both") {
        const summaries = await this.searchSummaries(ftsQuery);
        results.push(...summaries);
      }
    } catch (err) {
      // Source failure should not break recall — return whatever we have
      // Note: console.error won't appear in plugin logs; this is a last-resort fallback
    }

    return results.filter((s) => s.score >= minScore);
  }

  private async searchMessages(ftsQuery: string): Promise<RecallSnippet[]> {
    const sql = `SELECT m.message_id, m.role, substr(m.content, 1, 500) as content, m.created_at, rank as fts_rank FROM messages_fts f JOIN messages m ON m.rowid = f.rowid WHERE messages_fts MATCH '${ftsQuery}' AND m.role IN ('user', 'assistant') AND m.content NOT LIKE '{"type":"toolCall%' AND m.content NOT LIKE '{"type":"toolResult%' ORDER BY rank LIMIT ${this.config.maxResults}`;

    const rows = await this.db.query(sql);

    return rows.map((row) => {
      const score = normalizeFtsRank(Number(row.fts_rank) || 0);
      const messageId = String(row.message_id);
      const role = String(row.role || "unknown");
      const rawContent = String(row.content || "");
      const createdAt = row.created_at ? ` (${String(row.created_at)})` : "";

      // Clean the content — strip OpenClaw metadata, queued message headers,
      // and other envelope noise that clutters recall results
      const content = cleanLcmContent(rawContent);
      if (!content) return null;

      return {
        source: this.name,
        path: `lcm:messages/${messageId}`,
        text: `[${role}]${createdAt} ${content}`.trim(),
        score,
        reference: `lcm:messages/${messageId}`,
      };
    }).filter((s): s is RecallSnippet => s !== null);
  }

  private async searchSummaries(ftsQuery: string): Promise<RecallSnippet[]> {
    const sql = `SELECT f.summary_id, substr(f.content, 1, 500) as content, s.created_at, rank as fts_rank FROM summaries_fts f JOIN summaries s ON s.summary_id = f.summary_id WHERE summaries_fts MATCH '${ftsQuery}' ORDER BY rank LIMIT ${this.config.maxResults}`;

    const rows = await this.db.query(sql);

    return rows.map((row) => {
      const score = normalizeFtsRank(Number(row.fts_rank) || 0);
      const summaryId = String(row.summary_id);
      const content = String(row.content || "");
      const createdAt = row.created_at ? ` (${String(row.created_at)})` : "";

      return {
        source: this.name,
        path: `lcm:summaries/${summaryId}`,
        text: `[summary]${createdAt} ${content}`.trim(),
        score,
        reference: `lcm:summaries/${summaryId}`,
      };
    });
  }
}

/**
 * Clean LCM message content for recall display.
 * Strips OpenClaw metadata envelopes, queued message headers, and other noise
 * that makes recall results hard to read.
 */
function cleanLcmContent(content: string): string {
  let cleaned = content;

  // Strip "[Queued messages while agent was busy]" headers and dividers
  cleaned = cleaned.replace(/\[Queued messages while agent was busy\]\s*/g, "");
  cleaned = cleaned.replace(/---\s*Queued #\d+\s*/g, "");

  // Strip OpenClaw metadata blocks (JSON in code fences)
  cleaned = cleaned.replace(
    /(?:Conversation info|Sender|Replied message)\s*\(untrusted[^)]*\):\s*```json\s*\{[\s\S]*?\}\s*```/g,
    ""
  );

  // Strip "System:" lines (exec results, gateway notices)
  cleaned = cleaned.replace(/^System:.*$/gm, "");

  // Strip recalled context blocks (our own output)
  cleaned = cleaned.replace(/## 🧠 Recalled Context[\s\S]*?(?=\n\n[A-Z]|\n\n\[|$)/g, "");

  // Strip timestamp prefixes like [Tue 2026-03-24 01:46 MDT]
  cleaned = cleaned.replace(/^\[.*?\]\s*/gm, "");

  // Collapse whitespace
  cleaned = cleaned.replace(/\n{3,}/g, "\n\n").replace(/\s+/g, " ").trim();

  return cleaned;
}

/**
 * Normalize FTS5 rank to 0-1 score.
 * FTS5 rank is negative (lower = more relevant). Typical values range from
 * -1 to -15. We use a scaled logistic normalization so that rank -5 maps
 * to ~0.5 and rank -1 maps to ~0.83, making scores comparable to other sources.
 */
function normalizeFtsRank(rank: number): number {
  const absRank = Math.abs(rank);
  // Scale factor of 0.2 maps rank -5 to score ~0.5
  return 1 / (1 + absRank * 0.2);
}

/**
 * Sanitize a query string for FTS5 MATCH syntax.
 * Wraps each word in double quotes to prevent syntax errors from special chars.
 */
// Stop words to filter from FTS queries — these add noise to search results
const FTS_STOP_WORDS = new Set([
  "a", "an", "the", "and", "or", "but", "in", "on", "at", "to", "for",
  "of", "with", "by", "from", "is", "it", "was", "be", "are", "been",
  "have", "has", "had", "do", "does", "did", "will", "would", "could",
  "should", "may", "might", "can", "this", "that", "these", "those",
  "i", "you", "he", "she", "we", "they", "me", "him", "her", "us",
  "them", "my", "your", "his", "our", "their", "what", "which", "who",
  "when", "where", "why", "how", "not", "no", "if", "then", "so",
  "just", "also", "very", "too", "about", "up", "out", "off",
]);

function sanitizeFtsQuery(query: string): string {
  const words = query
    .replace(/[^a-zA-Z0-9\s]/g, "")
    .split(/\s+/)
    .filter((w) => w.length > 2 && !FTS_STOP_WORDS.has(w.toLowerCase()));

  if (words.length === 0) return "";

  // Take up to 8 key terms, use OR for inclusive matching
  return words.slice(0, 8).map((w) => `"${w}"`).join(" OR ");
}
