// @clawdactual/openclaw-recall/src/sources/memory-search.ts
/**
 * Memory Search source — uses OpenClaw's built-in memory_search API.
 *
 * This is the universal source that works for every OpenClaw agent,
 * regardless of what other memory systems they have installed.
 */

import type { RecallSnippet, RecallSource } from "../types.js";

export interface MemorySearchApi {
  search(query: string, maxResults?: number, minScore?: number): Promise<MemorySearchResult[]>;
}

export interface MemorySearchResult {
  path: string;
  text: string;
  score: number;
  lineStart?: number;
  lineEnd?: number;
}

export class MemorySearchSource implements RecallSource {
  readonly name = "memory-search";
  private api: MemorySearchApi;
  private maxResults: number;

  constructor(api: MemorySearchApi, maxResults: number = 5) {
    this.api = api;
    this.maxResults = maxResults;
  }

  async search(query: string, minScore: number): Promise<RecallSnippet[]> {
    try {
      const results = await this.api.search(query, this.maxResults, minScore);

      return results.map((r) => ({
        source: this.name,
        path: r.path,
        text: r.text.trim(),
        score: r.score,
        lineStart: r.lineStart,
        lineEnd: r.lineEnd,
        reference: formatReference(r),
      }));
    } catch (err) {
      // Source failure should not break recall — return empty
      console.error(`[recall] memory-search error:`, err instanceof Error ? err.message : err);
      return [];
    }
  }
}

function formatReference(result: MemorySearchResult): string {
  const path = result.path;
  if (result.lineStart && result.lineEnd) {
    return `${path}:${result.lineStart}-${result.lineEnd}`;
  }
  if (result.lineStart) {
    return `${path}:${result.lineStart}`;
  }
  return path;
}
