// @clawdactual/openclaw-recall/src/sources/qmd-search.ts
/**
 * QMD Search source — searches local markdown indexes via the `qmd` CLI.
 *
 * Uses `qmd search` (BM25 only, fast, no API calls) rather than `qmd query`
 * (which does expansion + reranking and takes 15+ seconds).
 */

import type { RecallSnippet, RecallSource } from "../types.js";

export interface QmdApi {
  search(query: string): Promise<unknown[]>;
}

interface QmdResult {
  docid?: string;
  score?: number;
  file?: string;
  title?: string;
  context?: string;
  snippet?: string;
}

export class QmdSearchSource implements RecallSource {
  readonly name = "qmd";
  private api: QmdApi;
  private maxResults: number;

  constructor(api: QmdApi, maxResults: number = 5) {
    this.api = api;
    this.maxResults = maxResults;
  }

  async search(query: string, minScore: number): Promise<RecallSnippet[]> {
    try {
      // BM25 uses implicit AND — multi-term queries often return empty.
      // Strategy: extract key terms, search each individually in parallel,
      // then merge/deduplicate. This gives OR semantics with proper ranking.
      const terms = extractKeyTerms(query).split(/\s+/).filter((t) => t.length > 0);
      if (terms.length === 0) return [];

      // Search all terms in parallel
      const searches = terms.slice(0, 4).map((term) => 
        this.api.search(term).catch(() => [] as unknown[])
      );
      const allResults = await Promise.all(searches);
      const raw = allResults.flat();

      if (raw.length === 0) return [];

      // Map, deduplicate by file path, keep highest score
      const seen = new Map<string, RecallSnippet>();
      for (const item of raw) {
        const mapped = this.mapResult(item as QmdResult);
        if (!mapped || mapped.score < minScore) continue;
        const existing = seen.get(mapped.path);
        if (!existing || mapped.score > existing.score) {
          // Boost score if found by multiple terms
          if (existing) {
            mapped.score = Math.min(1.0, mapped.score + 0.1);
          }
          seen.set(mapped.path, mapped);
        }
      }

      return [...seen.values()]
        .sort((a, b) => b.score - a.score)
        .slice(0, this.maxResults);
    } catch (err) {
      // Source failure should not break recall
      return [];
    }
  }

  private mapResult(item: QmdResult): RecallSnippet | null {
    if (!item.file || !item.snippet) return null;

    const path = cleanQmdPath(item.file);
    const text = cleanSnippet(item.snippet);
    const score = typeof item.score === "number" ? item.score : 0;

    if (!text) return null;

    return {
      source: this.name,
      path,
      text,
      score,
      reference: path,
    };
  }
}

// Common stop words to filter for BM25 queries
const STOP_WORDS = new Set([
  "a", "an", "the", "and", "or", "but", "in", "on", "at", "to", "for",
  "of", "with", "by", "from", "is", "it", "was", "be", "are", "been",
  "have", "has", "had", "do", "does", "did", "will", "would", "could",
  "should", "may", "might", "can", "this", "that", "these", "those",
  "i", "you", "he", "she", "we", "they", "me", "him", "her", "us",
  "them", "my", "your", "his", "our", "their", "what", "which", "who",
  "when", "where", "why", "how", "not", "no", "if", "then", "so",
  "just", "also", "very", "too", "about", "after", "before", "up",
  "down", "out", "off", "only", "own", "same", "other", "like",
  "get", "got", "go", "going", "make", "take", "want", "know",
  "think", "say", "said", "tell", "please", "hey", "ok", "yes", "yeah",
  "sure", "thing", "things", "way", "really", "right", "now", "let",
  "see", "look", "something", "anything", "new", "first", "last",
  "yet", "still", "never", "ever", "any", "all", "been", "being",
  "confirm", "wondering", "talking", "turned",
]);

/**
 * Extract key terms from a query for BM25 search.
 * Strips stop words and limits to the most important terms.
 */
function extractKeyTerms(query: string): string {
  const words = query
    .toLowerCase()
    .replace(/[^a-z0-9\s-]/g, " ")
    .split(/\s+/)
    .filter((w) => w.length > 2 && !STOP_WORDS.has(w));

  // BM25 uses implicit AND — too many terms = zero results.
  // Keep only the 2-3 most distinctive terms for best recall.
  return words.slice(0, 3).join(" ");
}

/**
 * Clean QMD file path: strip `qmd://` prefix and collection name.
 * e.g., "qmd://clawd/memory.md" → "memory.md"
 *       "qmd://memory/2026-03-20.md" → "memory/2026-03-20.md"
 *       "regular/path.md" → "regular/path.md"
 */
function cleanQmdPath(file: string): string {
  if (!file.startsWith("qmd://")) return file;

  // Strip "qmd://" prefix
  const withoutScheme = file.slice(6);

  // Strip collection name (first path segment)
  const slashIndex = withoutScheme.indexOf("/");
  if (slashIndex === -1) return withoutScheme;

  return withoutScheme.slice(slashIndex + 1);
}

/**
 * Clean QMD snippet: strip the `@@ ... @@` header line.
 * e.g., "@@ -141,4 @@ (140 before, 53 after)\n- content here..."
 *       → "- content here..."
 */
function cleanSnippet(snippet: string): string {
  const lines = snippet.split("\n");

  // If first line matches the @@ header pattern, remove it
  if (lines.length > 0 && /^@@\s.*@@/.test(lines[0])) {
    return lines.slice(1).join("\n").trim();
  }

  return snippet.trim();
}
