// @clawdactual/openclaw-recall/src/config.ts
/**
 * Recall plugin configuration — resolved from plugin config + defaults.
 */

export interface RecallSourceConfig {
  memorySearch: {
    enabled: boolean;
    maxResults: number;
  };
  qmd: {
    enabled: boolean;
    command: string;
    maxResults: number;
  };
  lcm: {
    enabled: boolean;
    maxResults: number;
    scope: "messages" | "summaries" | "both";
  };
  engram: {
    enabled: boolean;
    command: string;
    maxResults: number;
    excludeModels: string[];
  };
}

export interface RecallConfig {
  enabled: boolean;
  skipMemorySearchIfLancedb: boolean;
  maxSnippets: number;
  minRelevanceScore: number;
  skipShortMessages: {
    enabled: boolean;
    minChars: number;
  };
  skipHeartbeats: boolean;
  sources: RecallSourceConfig;
  /**
   * Files to exclude from recall results — these are already in context.
   * Paths are matched against the end of result paths (case-insensitive).
   * Default: OpenClaw's standard bootstrap files.
   */
  excludeFiles: string[];
  /**
   * Glob-style patterns to exclude from recall results.
   * Supports * wildcards. Matched against full path (case-insensitive).
   * Example: ["memory/2026-*.md"] to exclude loaded daily files.
   */
  excludePatterns: string[];
  debug: {
    enabled: boolean;
    logPath: string;
  };
}

/**
 * Default bootstrap files that OpenClaw injects into Project Context.
 * No point recalling content the agent already has in its context window.
 */
const DEFAULT_EXCLUDE_FILES = [
  "AGENTS.md",
  "SOUL.md",
  "TOOLS.md",
  "IDENTITY.md",
  "USER.md",
  "HEARTBEAT.md",
  "BOOTSTRAP.md",
  "MEMORY.md",
  "PERSONALITY.md",
];

const DEFAULTS: RecallConfig = {
  enabled: true,
  skipMemorySearchIfLancedb: true,
  maxSnippets: 5,
  minRelevanceScore: 0.3,
  skipShortMessages: {
    enabled: true,
    minChars: 40,
  },
  skipHeartbeats: true,
  excludeFiles: DEFAULT_EXCLUDE_FILES,
  excludePatterns: [
    "**/node_modules/**",
    "**/dist/**",
    "**/build/**",
    "**/lib/**",
    "**/vendor/**",
    "**/changelog.md",
    "**/CHANGELOG.md",
    "**/license.md",
    "**/LICENSE.md",
    "**/contributing.md",
    "**/CONTRIBUTING.md",
    "**/code-of-conduct.md",
    "**/CODE_OF_CONDUCT.md",
    "**/security.md",
    "**/SECURITY.md",

  ],
  sources: {
    memorySearch: {
      enabled: true,
      maxResults: 5,
    },
    qmd: {
      enabled: false,
      command: "qmd search",
      maxResults: 5,
    },
    lcm: {
      enabled: false,
      maxResults: 3,
      scope: "summaries",
    },
    engram: {
      enabled: false,
      command: "engram",
      maxResults: 5,
      excludeModels: [],
    },
  },
  debug: {
    enabled: false,
    logPath: ".recall/debug.log",
  },
};

/**
 * Resolve recall config from raw plugin config, applying defaults.
 */
export function resolveRecallConfig(raw: Record<string, unknown> | undefined): RecallConfig {
  if (!raw) return { ...DEFAULTS };

  // Support both old name (skipIfLancedb) and new name (skipMemorySearchIfLancedb)
  const skipMemorySearchIfLancedb =
    typeof raw.skipMemorySearchIfLancedb === "boolean"
      ? raw.skipMemorySearchIfLancedb
      : typeof raw.skipIfLancedb === "boolean"
        ? raw.skipIfLancedb
        : DEFAULTS.skipMemorySearchIfLancedb;

  const config: RecallConfig = {
    enabled: typeof raw.enabled === "boolean" ? raw.enabled : DEFAULTS.enabled,
    skipMemorySearchIfLancedb,
    maxSnippets:
      typeof raw.maxSnippets === "number"
        ? Math.max(1, Math.min(20, Math.floor(raw.maxSnippets)))
        : DEFAULTS.maxSnippets,
    minRelevanceScore:
      typeof raw.minRelevanceScore === "number"
        ? Math.max(0, Math.min(1, raw.minRelevanceScore))
        : DEFAULTS.minRelevanceScore,
    skipShortMessages: resolveSkipShortMessages(raw.skipShortMessages),
    skipHeartbeats:
      typeof raw.skipHeartbeats === "boolean" ? raw.skipHeartbeats : DEFAULTS.skipHeartbeats,
    excludeFiles: resolveExcludeFiles(raw.excludeFiles),
    excludePatterns: resolveExcludePatterns(raw.excludePatterns),
    sources: resolveSources(raw.sources),
    debug: resolveDebug(raw.debug),
  };

  return config;
}

function resolveSkipShortMessages(raw: unknown): RecallConfig["skipShortMessages"] {
  if (!raw || typeof raw !== "object") return { ...DEFAULTS.skipShortMessages };
  const obj = raw as Record<string, unknown>;
  return {
    enabled: typeof obj.enabled === "boolean" ? obj.enabled : DEFAULTS.skipShortMessages.enabled,
    minChars:
      typeof obj.minChars === "number"
        ? Math.max(1, Math.floor(obj.minChars))
        : DEFAULTS.skipShortMessages.minChars,
  };
}

function resolveSources(raw: unknown): RecallSourceConfig {
  if (!raw || typeof raw !== "object") return { ...DEFAULTS.sources };
  const obj = raw as Record<string, unknown>;

  return {
    memorySearch: resolveMemorySearchSource(obj.memorySearch),
    qmd: resolveQmdSource(obj.qmd),
    lcm: resolveLcmSource(obj.lcm),
    engram: resolveEngramSource(obj.engram),
  };
}

function resolveMemorySearchSource(raw: unknown): RecallSourceConfig["memorySearch"] {
  if (!raw || typeof raw !== "object") return { ...DEFAULTS.sources.memorySearch };
  const obj = raw as Record<string, unknown>;
  return {
    enabled: typeof obj.enabled === "boolean" ? obj.enabled : DEFAULTS.sources.memorySearch.enabled,
    maxResults:
      typeof obj.maxResults === "number"
        ? Math.max(1, Math.floor(obj.maxResults))
        : DEFAULTS.sources.memorySearch.maxResults,
  };
}

function resolveQmdSource(raw: unknown): RecallSourceConfig["qmd"] {
  if (!raw || typeof raw !== "object") return { ...DEFAULTS.sources.qmd };
  const obj = raw as Record<string, unknown>;
  return {
    enabled: typeof obj.enabled === "boolean" ? obj.enabled : DEFAULTS.sources.qmd.enabled,
    command: typeof obj.command === "string" ? obj.command.trim() : DEFAULTS.sources.qmd.command,
    maxResults:
      typeof obj.maxResults === "number"
        ? Math.max(1, Math.floor(obj.maxResults))
        : DEFAULTS.sources.qmd.maxResults,
  };
}

function resolveLcmSource(raw: unknown): RecallSourceConfig["lcm"] {
  if (!raw || typeof raw !== "object") return { ...DEFAULTS.sources.lcm };
  const obj = raw as Record<string, unknown>;
  const scope = obj.scope;
  return {
    enabled: typeof obj.enabled === "boolean" ? obj.enabled : DEFAULTS.sources.lcm.enabled,
    maxResults:
      typeof obj.maxResults === "number"
        ? Math.max(1, Math.floor(obj.maxResults))
        : DEFAULTS.sources.lcm.maxResults,
    scope:
      scope === "messages" || scope === "summaries" || scope === "both"
        ? scope
        : DEFAULTS.sources.lcm.scope,
  };
}

function resolveEngramSource(raw: unknown): RecallSourceConfig["engram"] {
  if (!raw || typeof raw !== "object") return { ...DEFAULTS.sources.engram };
  const obj = raw as Record<string, unknown>;
  return {
    enabled: typeof obj.enabled === "boolean" ? obj.enabled : DEFAULTS.sources.engram.enabled,
    command: typeof obj.command === "string" ? obj.command.trim() : DEFAULTS.sources.engram.command,
    maxResults:
      typeof obj.maxResults === "number"
        ? Math.max(1, Math.floor(obj.maxResults))
        : DEFAULTS.sources.engram.maxResults,
    excludeModels: Array.isArray(obj.excludeModels)
      ? obj.excludeModels.filter((item): item is string => typeof item === "string")
      : DEFAULTS.sources.engram.excludeModels,
  };
}

function resolveExcludePatterns(raw: unknown): string[] {
  // Always start with defaults — user patterns are additive, not replacement.
  // This prevents the common mistake of providing a custom list that silently
  // drops important defaults (e.g., node_modules, dist).
  const defaults = new Set(DEFAULTS.excludePatterns.map((p) => p.toLowerCase()));
  const result = [...DEFAULTS.excludePatterns];

  if (Array.isArray(raw)) {
    for (const item of raw) {
      if (typeof item === "string") {
        const trimmed = item.trim();
        if (trimmed.length > 0 && !defaults.has(trimmed.toLowerCase())) {
          result.push(trimmed);
        }
      }
    }
  }

  return result;
}

function resolveExcludeFiles(raw: unknown): string[] {
  if (Array.isArray(raw)) {
    const files = raw
      .filter((item): item is string => typeof item === "string")
      .map((f) => f.trim())
      .filter((f) => f.length > 0);
    // If user provides an explicit list, use it (they may want to override defaults)
    return files.length > 0 ? files : DEFAULT_EXCLUDE_FILES;
  }
  // No override — use defaults
  return DEFAULT_EXCLUDE_FILES;
}

function resolveDebug(raw: unknown): RecallConfig["debug"] {
  if (!raw || typeof raw !== "object") return { ...DEFAULTS.debug };
  const obj = raw as Record<string, unknown>;
  return {
    enabled: typeof obj.enabled === "boolean" ? obj.enabled : DEFAULTS.debug.enabled,
    logPath: typeof obj.logPath === "string" ? obj.logPath.trim() : DEFAULTS.debug.logPath,
  };
}
