// @clawdactual/openclaw-recall/src/query-extractor.ts
/**
 * Query extraction — transforms user messages into focused search queries.
 *
 * Strategy adapts based on message length:
 * - Short (<500 chars): search verbatim
 * - Medium (500-2000 chars): first sentence + key phrases
 * - Long (>2000 chars): segment into chunks, extract key phrase per chunk
 *
 * Key improvements (v0.3):
 * - Multi-word entity extraction (adjacent capitalized words → single phrase)
 * - CamelCase splitting for additional search variants
 * - Co-occurrence scoring (RAKE-inspired degree/frequency)
 * - Abbreviation expansion via configurable domain dictionary
 * - Casual filter respects entity presence (won't skip "How's OathKeeper going?")
 */

// Common English stop words to filter out during key phrase extraction
const STOP_WORDS = new Set([
  "a", "an", "the", "and", "or", "but", "in", "on", "at", "to", "for",
  "of", "with", "by", "from", "is", "it", "its", "was", "be", "are",
  "been", "being", "have", "has", "had", "do", "does", "did", "will",
  "would", "could", "should", "may", "might", "shall", "can", "need",
  "this", "that", "these", "those", "i", "you", "he", "she", "we",
  "they", "me", "him", "her", "us", "them", "my", "your", "his",
  "our", "their", "what", "which", "who", "whom", "when", "where",
  "why", "how", "not", "no", "nor", "if", "then", "else", "so",
  "just", "also", "very", "too", "quite", "rather", "about", "above",
  "after", "before", "between", "under", "over", "into", "through",
  "during", "each", "few", "more", "most", "some", "any", "all",
  "both", "here", "there", "again", "once", "than", "as", "up",
  "down", "out", "off", "only", "own", "same", "such", "other",
  "like", "get", "got", "go", "going", "make", "made", "take",
  "want", "know", "think", "say", "said", "tell", "please", "hey",
  "ok", "okay", "thanks", "hi", "hello", "yes", "yeah", "sure",
  "thing", "things", "way", "time", "lot", "much", "many", "well",
  "really", "right", "now", "let", "see", "look", "something",
  "everything", "nothing", "anything", "new", "first", "last",
]);

/**
 * Default abbreviation/synonym expansion dictionary.
 * Users can extend this via config. Keys are lowercase.
 * Values are expansion strings added to queries for broader recall.
 */
const DEFAULT_ABBREVIATIONS: ReadonlyMap<string, string> = new Map([
  ["pr", "pull request"],
  ["prs", "pull requests"],
  ["e2e", "end-to-end"],
  ["ci", "continuous integration"],
  ["cd", "continuous deployment"],
  ["api", "application programming interface"],
  ["db", "database"],
  ["ui", "user interface"],
  ["ux", "user experience"],
  ["tdd", "test driven development"],
  ["sdk", "software development kit"],
  ["cli", "command line interface"],
  ["mcp", "model context protocol"],
  ["rag", "retrieval augmented generation"],
  ["llm", "large language model"],
  ["ssr", "server side rendering"],
  ["jwt", "json web token"],
  ["orm", "object relational mapping"],
  ["crud", "create read update delete"],
  ["dto", "data transfer object"],
  ["env", "environment"],
  ["config", "configuration"],
  ["deps", "dependencies"],
  ["repo", "repository"],
  ["repos", "repositories"],
  ["infra", "infrastructure"],
  ["auth", "authentication"],
  ["authn", "authentication"],
  ["authz", "authorization"],
]);

// Mutable store for user-provided abbreviations (set at init time)
let customAbbreviations: Map<string, string> = new Map();

/**
 * Set custom abbreviation expansions (called from plugin init).
 * These are merged with defaults; custom entries override defaults.
 */
export function setCustomAbbreviations(abbrevs: Record<string, string>): void {
  customAbbreviations = new Map(Object.entries(abbrevs).map(([k, v]) => [k.toLowerCase(), v]));
}

/**
 * Look up an abbreviation expansion. Custom overrides default.
 */
function expandAbbreviation(term: string): string | undefined {
  const lower = term.toLowerCase();
  return customAbbreviations.get(lower) ?? DEFAULT_ABBREVIATIONS.get(lower);
}

/**
 * Clean a raw message — strip system metadata, reply tags, channel noise.
 *
 * The `before_agent_start` hook receives the full prompt including OpenClaw
 * envelope metadata (conversation info, sender info, reply context, system
 * messages). We need to extract just the human's actual message.
 */
/**
 * Exported for use by the recall engine (short message / casual detection).
 * Named differently to make the export purpose clear.
 */
export function cleanMessageForRecall(text: string): string {
  return cleanMessage(text);
}

function cleanMessage(text: string): string {
  let cleaned = text;

  // Strip OpenClaw reply tags
  cleaned = cleaned.replace(/\[\[\s*reply_to[^\]]*\]\]/g, "");

  // Strip OpenClaw inbound metadata blocks (JSON in code fences with untrusted metadata headers)
  cleaned = cleaned.replace(/(?:Conversation info|Sender|Replied message)\s*\(untrusted[^)]*\):\s*```json\s*\{[\s\S]*?\}\s*```/g, "");

  // Strip our own recalled context block (avoid searching for our own output)
  cleaned = cleaned.replace(/## 🧠 Recalled Context[\s\S]*?(?=\n(?:Conversation info|Sender|System:|[A-Z])|\n\n[A-Z]|$)/g, "");

  // Strip thread history blocks (Slack/Discord)
  cleaned = cleaned.replace(/\[Thread history - for context\][\s\S]*?(?=\n\n|\n[A-Z]|$)/g, "");

  // Strip "System:" prefixed lines (system events, exec results, gateway restart notices)
  cleaned = cleaned.replace(/^System:.*$/gm, "");

  // Strip common system prefixes like [timestamp] or [channel info]
  cleaned = cleaned.replace(/^\[.*?\]\s*/gm, "");

  // Strip markdown formatting but keep content
  cleaned = cleaned.replace(/```[^`]*```/g, ""); // code blocks
  cleaned = cleaned.replace(/`[^`]+`/g, (match) => match.slice(1, -1)); // inline code
  cleaned = cleaned.replace(/[*_~]+/g, ""); // bold/italic/strikethrough

  // Collapse whitespace
  cleaned = cleaned.replace(/\s+/g, " ").trim();

  return cleaned;
}

/**
 * Extract the first 1-2 sentences from text.
 */
function extractFirstSentences(text: string, count: number = 2): string {
  // Split on sentence boundaries (., !, ? followed by space or end)
  const sentences = text.split(/(?<=[.!?])\s+/);
  return sentences.slice(0, count).join(" ").trim();
}

/**
 * Extract key phrases using RAKE-inspired co-occurrence scoring.
 *
 * RAKE scores words by degree(word)/frequency(word), where "degree" counts
 * how many other content words co-occur with this word in candidate phrases.
 * Words that appear alongside many other important words score higher than
 * words that just appear frequently in isolation.
 *
 * Returns 2-3 focused query phrases.
 */
function extractKeyPhrases(text: string, maxPhrases: number = 3): string[] {
  // Step 1: Split text into candidate phrases at stop-word boundaries (RAKE core)
  const candidatePhrases = text
    .toLowerCase()
    .replace(/[^a-z0-9\s-]/g, " ")
    .split(/\s+/)
    .reduce<string[][]>(
      (acc, word) => {
        if (STOP_WORDS.has(word) || word.length <= 2) {
          // Stop word boundary — start new candidate phrase
          if (acc[acc.length - 1].length > 0) acc.push([]);
        } else {
          acc[acc.length - 1].push(word);
        }
        return acc;
      },
      [[]],
    )
    .filter((phrase) => phrase.length > 0);

  // Step 2: Calculate word scores using degree/frequency
  const freq = new Map<string, number>();
  const degree = new Map<string, number>();

  for (const phrase of candidatePhrases) {
    const phraseLen = phrase.length;
    for (const word of phrase) {
      freq.set(word, (freq.get(word) || 0) + 1);
      // Degree: sum of phrase lengths containing this word (minus 1 for self)
      // This counts co-occurrence with other content words
      degree.set(word, (degree.get(word) || 0) + phraseLen - 1);
    }
  }

  // Word score = (degree + frequency) / frequency
  // This is the standard RAKE formula: words that co-occur with many
  // other content words get boosted relative to frequency-only scoring
  const wordScores = new Map<string, number>();
  for (const [word, f] of freq) {
    const d = degree.get(word) || 0;
    wordScores.set(word, (d + f) / f);
  }

  // Step 3: Score candidate phrases by summing word scores
  // Also apply position weight (earlier phrases rank higher)
  const totalPhrases = candidatePhrases.length;
  const scoredPhrases: Array<{ phrase: string; score: number }> = [];

  for (let i = 0; i < candidatePhrases.length; i++) {
    const phraseWords = candidatePhrases[i];
    const phraseScore = phraseWords.reduce(
      (sum, w) => sum + (wordScores.get(w) || 0),
      0,
    );
    // Position decay: first phrase gets 1.0, last gets 0.6
    const posWeight = 1.0 - (i / Math.max(totalPhrases, 1)) * 0.4;
    scoredPhrases.push({
      phrase: phraseWords.join(" "),
      score: phraseScore * posWeight,
    });
  }

  // Sort by score descending
  scoredPhrases.sort((a, b) => b.score - a.score);

  // Step 4: Select top phrases, avoiding redundancy
  const selected: string[] = [];
  for (const { phrase } of scoredPhrases) {
    if (selected.length >= maxPhrases) break;
    // Skip if this phrase is a subset of an already-selected one
    if (selected.some((s) => s.includes(phrase) || phrase.includes(s))) continue;
    // Skip single very short words
    if (phrase.length < 4) continue;
    selected.push(phrase);
  }

  // Step 5: If we didn't get enough phrases, fall back to top individual words
  if (selected.length < maxPhrases) {
    const topWords = [...wordScores.entries()]
      .sort((a, b) => b[1] - a[1])
      .map(([word]) => word);
    for (const word of topWords) {
      if (selected.length >= maxPhrases) break;
      if (!selected.some((p) => p.includes(word))) {
        selected.push(word);
      }
    }
  }

  // Step 6: Expand abbreviations found in selected phrases
  const expanded: string[] = [];
  for (const phrase of selected) {
    expanded.push(phrase);
    // Check each word in the phrase for abbreviation expansion
    for (const word of phrase.split(/\s+/)) {
      const expansion = expandAbbreviation(word);
      if (expansion && !expanded.includes(expansion)) {
        expanded.push(expansion);
      }
    }
  }

  return expanded.slice(0, maxPhrases + 2); // Allow a couple extra from expansions
}

// extractNgramPhrases removed in v0.3 — replaced by RAKE-based phrase extraction
// which naturally produces n-gram candidates from stop-word-delimited segments.

/**
 * Segment a long message into logical chunks.
 */
function segmentMessage(text: string): string[] {
  // Split on double newlines (paragraph breaks)
  let segments = text.split(/\n\s*\n/).filter((s) => s.trim().length > 0);

  // If still too few segments, try bullet points
  if (segments.length <= 1) {
    segments = text.split(/\n\s*[•\-*]\s+/).filter((s) => s.trim().length > 0);
  }

  // If still one big block, split by sentences into ~3 groups
  if (segments.length <= 1) {
    const sentences = text.split(/(?<=[.!?])\s+/).filter((s) => s.trim().length > 0);
    const groupSize = Math.ceil(sentences.length / 3);
    segments = [];
    for (let i = 0; i < sentences.length; i += groupSize) {
      segments.push(sentences.slice(i, i + groupSize).join(" "));
    }
  }

  return segments.map((s) => s.trim()).filter((s) => s.length > 0);
}

/**
 * Extract proper nouns, multi-word entities, and capitalized terms from text.
 * These are high-signal terms (project names, people, tools) that should
 * always be included in search queries regardless of position in message.
 *
 * Improvements over v0.2:
 * - Adjacent capitalized words are merged into multi-word entities
 *   ("Tom Merkle" → one phrase, "Mental Model" → one phrase)
 * - CamelCase terms also produce a split variant
 *   ("OathKeeper" → also yields "Oath Keeper" for broader matching)
 * - Abbreviation expansions are appended where matched
 */
function extractProperNouns(text: string): string[] {
  const entities: string[] = [];
  const words = text.split(/\s+/);

  // Skip common sentence-starting / function words
  const commonStarters = new Set([
    "The", "This", "That", "These", "Those", "It", "Its", "I",
    "We", "You", "He", "She", "They", "My", "Your", "Our",
    "How", "What", "When", "Where", "Why", "Who", "Which",
    "Also", "And", "But", "Or", "So", "If", "For", "Not",
    "Sure", "Ok", "Yes", "No", "Hey", "Hi", "Just", "Well",
    "Now", "Then", "Here", "There", "Let", "Do", "Does", "Did",
    "Can", "Could", "Would", "Should", "Will", "May", "Might",
  ]);

  function isProperNoun(word: string, index: number): boolean {
    if (!word || word.length < 2) return false;

    const isCapitalized = /^[A-Z]/.test(word);
    const isCamelCase = /[a-z][A-Z]/.test(word);
    const isAllCaps = /^[A-Z]{2,}$/.test(word) && word.length <= 10;
    const isSentenceStart = index === 0 || /[.!?]\s*$/.test(words[index - 1] || "");

    // Check against common starters — also check the base word before
    // apostrophe/contraction was stripped (e.g., "How's" → "Hows" should
    // still match "How" in commonStarters)
    const rawWord = words[index]?.replace(/[^a-zA-Z]/g, "") ?? word;
    const baseBeforeContraction = rawWord.replace(/'s$|'t$|'re$|'ve$|'ll$|'d$/i, "")
      || word.replace(/s$/, ""); // "Hows" → "How"
    const isCommonStarter =
      commonStarters.has(word) ||
      commonStarters.has(baseBeforeContraction) ||
      // Also catch apostrophe-stripped contractions: "Hows" (from "How's"), "Whats", etc.
      commonStarters.has(word.replace(/s$/i, ""));

    if (isCamelCase || isAllCaps) return true;
    if (isCapitalized && !isSentenceStart && !isCommonStarter) return true;
    if (isCapitalized && isSentenceStart && !isCommonStarter && !STOP_WORDS.has(word.toLowerCase())) return true;
    return false;
  }

  // Pass 1: Merge adjacent proper nouns into multi-word entities
  let i = 0;
  while (i < words.length) {
    const cleaned = words[i].replace(/[^a-zA-Z0-9-]/g, "");
    if (isProperNoun(cleaned, i)) {
      // Collect consecutive proper nouns
      const entityParts = [cleaned];
      let j = i + 1;
      while (j < words.length) {
        const nextCleaned = words[j].replace(/[^a-zA-Z0-9-]/g, "");
        // Allow a proper noun chain — but also allow a single lowercase connector
        // e.g., "Fleet Manager" yes, "Tom of Merkle" no
        if (isProperNoun(nextCleaned, j)) {
          entityParts.push(nextCleaned);
          j++;
        } else {
          break;
        }
      }

      // Add the full multi-word entity
      entities.push(entityParts.join(" "));
      // Also add individual words if multi-word (for partial matches)
      if (entityParts.length > 1) {
        for (const part of entityParts) {
          if (part.length >= 3) entities.push(part);
        }
      }
      i = j;
    } else {
      i++;
    }
  }

  // Pass 2: CamelCase splitting — "OathKeeper" → also "Oath Keeper"
  const camelEntities: string[] = [];
  for (const entity of entities) {
    if (/[a-z][A-Z]/.test(entity)) {
      const split = entity.replace(/([a-z])([A-Z])/g, "$1 $2");
      if (split !== entity && split.length >= 3) {
        camelEntities.push(split);
      }
    }
  }
  entities.push(...camelEntities);

  // Pass 3: Abbreviation expansion for short ALL-CAPS terms
  const abbreviationExpansions: string[] = [];
  for (const entity of entities) {
    const expansion = expandAbbreviation(entity);
    if (expansion) {
      abbreviationExpansions.push(expansion);
    }
  }
  entities.push(...abbreviationExpansions);

  // Deduplicate, preserving order
  return [...new Set(entities)];
}

/**
 * Check whether a cleaned message contains substantive entities
 * (proper nouns, domain terms, abbreviations). Used by the casual filter
 * to avoid skipping messages like "How's OathKeeper going?"
 */
export function hasSubstantiveEntities(text: string): boolean {
  const nouns = extractProperNouns(text);
  if (nouns.length > 0) return true;

  // Also check for known abbreviations in the text
  const words = text.toLowerCase().split(/\s+/);
  for (const word of words) {
    const cleaned = word.replace(/[^a-z0-9]/g, "");
    if (expandAbbreviation(cleaned)) return true;
  }
  return false;
}

export interface ExtractedQueries {
  queries: string[];
  strategy: "verbatim" | "first-sentence+keyphrases" | "segmented" | "proper-nouns";
}

/**
 * Extract search queries from a user message.
 */
export function extractQueries(rawMessage: string): ExtractedQueries {
  const cleaned = cleanMessage(rawMessage);

  if (!cleaned || cleaned.length === 0) {
    return { queries: [], strategy: "verbatim" };
  }

  // Always extract proper nouns — these are high-signal regardless of message length
  const properNouns = extractProperNouns(cleaned);

  // Short messages: search verbatim, plus proper nouns as separate query if found
  if (cleaned.length < 500) {
    const queries = [cleaned];
    if (properNouns.length > 0) {
      const nounQuery = properNouns.join(" ");
      if (nounQuery !== cleaned) {
        queries.push(nounQuery);
      }
    }
    return { queries, strategy: properNouns.length > 0 ? "proper-nouns" : "verbatim" };
  }

  // Medium messages: first sentence + key phrases + proper nouns
  if (cleaned.length <= 2000) {
    const queries: string[] = [];

    // Proper nouns first — they're the most targeted query
    if (properNouns.length > 0) {
      queries.push(properNouns.join(" "));
    }

    const firstSentence = extractFirstSentences(cleaned, 1);
    if (firstSentence.length > 10) {
      queries.push(firstSentence);
    }

    const keyPhrases = extractKeyPhrases(cleaned, 2);
    for (const phrase of keyPhrases) {
      if (!queries.some((q) => q.includes(phrase))) {
        queries.push(phrase);
      }
    }

    return {
      queries: queries.length > 0 ? queries : [cleaned.slice(0, 500)],
      strategy: "first-sentence+keyphrases",
    };
  }

  // Long messages: proper nouns first, then segment and extract per chunk
  const segments = segmentMessage(cleaned);
  const queries: string[] = [];

  // Proper nouns as first query for long messages too
  if (properNouns.length > 0) {
    queries.push(properNouns.join(" "));
  }

  for (const segment of segments.slice(0, 5)) {
    const phrases = extractKeyPhrases(segment, 1);
    if (phrases.length > 0) {
      queries.push(phrases[0]);
    } else {
      // Fallback: first sentence of segment
      const first = extractFirstSentences(segment, 1);
      if (first.length > 10) {
        queries.push(first);
      }
    }
  }

  return {
    queries: queries.length > 0 ? queries : [extractFirstSentences(cleaned, 2)],
    strategy: "segmented",
  };
}
