// @clawdactual/openclaw-recall/src/recall-engine.ts
/**
 * RecallEngine — the core orchestrator.
 *
 * Takes a user message, extracts queries, searches all configured sources
 * in parallel, ranks/deduplicates results, and formats the recall block.
 */

import type { RecallConfig } from "./config.js";
import { extractQueries, cleanMessageForRecall, hasSubstantiveEntities } from "./query-extractor.js";
import type { RecallResult, RecallSnippet, RecallSource } from "./types.js";

export class RecallEngine {
  private sources: RecallSource[] = [];
  private config: RecallConfig;
  private log: {
    info: (msg: string) => void;
    debug: (msg: string) => void;
    warn: (msg: string) => void;
    error: (msg: string) => void;
  };

  constructor(
    config: RecallConfig,
    log?: {
      info: (msg: string) => void;
      debug: (msg: string) => void;
      warn: (msg: string) => void;
      error: (msg: string) => void;
    },
  ) {
    this.config = config;
    this.log = log ?? {
      info: (msg) => console.log(`[recall] ${msg}`),
      debug: (msg) => console.debug(`[recall] ${msg}`),
      warn: (msg) => console.warn(`[recall] ${msg}`),
      error: (msg) => console.error(`[recall] ${msg}`),
    };
  }

  /**
   * Register a search source.
   */
  addSource(source: RecallSource): void {
    this.sources.push(source);
    this.log.info(`Registered source: ${source.name}`);
  }

  /**
   * Run recall for a user message.
   *
   * @param message - The raw user message text
   * @returns The assembled recall result, or null if nothing relevant found
   */
  async recall(message: string): Promise<RecallResult | null> {
    const startTime = Date.now();

    // Check skip conditions
    if (!this.config.enabled) {
      return null;
    }

    if (this.sources.length === 0) {
      return null;
    }

    const trimmed = message.trim();
    if (!trimmed) {
      return null;
    }

    if (this.config.skipHeartbeats && isHeartbeatMessage(trimmed)) {
      this.log.debug("Skipping recall: heartbeat message");
      return null;
    }

    // Clean the message (strip metadata envelope, recalled context, system lines)
    // before checking length and casual patterns — raw message includes
    // OpenClaw metadata that inflates the character count.
    const cleaned = cleanMessageForRecall(trimmed);
    if (!cleaned) {
      return null;
    }

    if (this.config.skipShortMessages.enabled && cleaned.length < this.config.skipShortMessages.minChars) {
      // Override: if the short message contains substantive entities, recall anyway
      // (e.g., "OathKeeper?" at 12 chars should still trigger recall)
      if (!hasSubstantiveEntities(cleaned)) {
        this.log.debug(`Skipping recall: message too short (${cleaned.length} < ${this.config.skipShortMessages.minChars})`);
        return null;
      }
    }

    if (isCasualMessage(cleaned)) {
      this.log.debug(`Skipping recall: casual message ("${cleaned.slice(0, 30)}")`);
      return null;
    }

    // Extract search queries
    const { queries, strategy } = extractQueries(trimmed);
    if (queries.length === 0) {
      return null;
    }

    this.log.debug(
      `Extracting queries (strategy=${strategy}): ${queries.map((q) => `"${q.slice(0, 60)}"`).join(", ")}`,
    );

    // Search all sources in parallel for all queries
    const allSnippets = await this.searchAllSources(queries);

    if (allSnippets.length === 0) {
      this.log.debug("No results from any source");
      return null;
    }

    // Filter out files already in context (bootstrap files)
    const filtered = this.filterExcludedFiles(allSnippets);
    if (filtered.length === 0) {
      this.log.debug("All results filtered (already in context)");
      return null;
    }

    // Rank, deduplicate, and take top N
    const ranked = this.rankAndDeduplicate(filtered);
    const fitted = this.selectTopSnippets(ranked);

    if (fitted.length === 0) {
      return null;
    }

    // Format the recall block
    const durationMs = Date.now() - startTime;
    const block = this.formatBlock(fitted, durationMs);
    const estimatedTokens = Math.ceil(block.length / 4); // rough: 1 token ≈ 4 chars

    this.log.info(
      `Recalled ${fitted.length} snippets (${estimatedTokens} est. tokens, ${durationMs}ms, strategy=${strategy})`,
    );

    return {
      block,
      estimatedTokens,
      snippetCount: fitted.length,
      totalResults: allSnippets.length,
      queryStrategy: strategy,
      durationMs,
    };
  }

  /**
   * Search all registered sources with all queries in parallel.
   */
  private async searchAllSources(queries: string[]): Promise<RecallSnippet[]> {
    const searchJobs: Promise<RecallSnippet[]>[] = [];

    for (const source of this.sources) {
      for (const query of queries) {
        searchJobs.push(
          source.search(query, this.config.minRelevanceScore).catch((err) => {
            this.log.warn(`Source ${source.name} failed for query "${query.slice(0, 50)}": ${err}`);
            return [] as RecallSnippet[];
          }),
        );
      }
    }

    const results = await Promise.all(searchJobs);
    return results.flat();
  }

  /**
   * Filter out snippets from files that are already loaded in context.
   * No point recalling what the agent already has.
   */
  private filterExcludedFiles(snippets: RecallSnippet[]): RecallSnippet[] {
    const hasFiles = this.config.excludeFiles.length > 0;
    const hasPatterns = this.config.excludePatterns.length > 0;

    if (!hasFiles && !hasPatterns) return snippets;

    const excludeFiles = this.config.excludeFiles.map((f) => f.toLowerCase());
    const excludeRegexes = hasPatterns
      ? this.config.excludePatterns.map((p) => globToRegex(p))
      : [];

    return snippets.filter((snippet) => {
      const path = snippet.path.toLowerCase();

      // Check exact file matches (suffix-based)
      if (hasFiles) {
        const matchesFile = excludeFiles.some(
          (f) => path.endsWith(f) || path.endsWith("/" + f),
        );
        if (matchesFile) return false;
      }

      // Check glob patterns
      if (hasPatterns) {
        const matchesPattern = excludeRegexes.some((re) => re.test(path));
        if (matchesPattern) return false;
      }

      return true;
    });
  }

  /**
   * Rank and deduplicate snippets from all sources/queries.
   */
  private rankAndDeduplicate(snippets: RecallSnippet[]): RecallSnippet[] {
    // Group by normalized reference for deduplication.
    // Same file from different sources/collections should merge.
    const groups = new Map<string, RecallSnippet[]>();

    for (const snippet of snippets) {
      const key = normalizeDedupeKey(snippet);
      const group = groups.get(key) || [];
      group.push(snippet);
      groups.set(key, group);
    }

    // For each group, keep the highest-scored snippet
    // Cross-source/query boost: if same content appears multiple times, boost score
    const deduped: RecallSnippet[] = [];

    for (const [_key, group] of groups) {
      // Sort by score descending
      group.sort((a, b) => b.score - a.score);
      const best = { ...group[0] };

      // Cross-hit boost: 10% per additional hit (from different sources or queries)
      const uniqueSources = new Set(group.map((s) => s.source));
      const crossBoost = Math.min(0.3, (group.length - 1) * 0.1);
      best.score = Math.min(1.0, best.score + crossBoost);

      deduped.push(best);
    }

    // Sort by final score descending
    deduped.sort((a, b) => b.score - a.score);

    return deduped;
  }

  /**
   * Select top N snippets by score, truncating text for cleaner context.
   * maxSnippets is the sole control for recall volume.
   */
  private selectTopSnippets(snippets: RecallSnippet[]): RecallSnippet[] {
    const selected: RecallSnippet[] = [];

    for (const snippet of snippets) {
      if (selected.length >= this.config.maxSnippets) break;

      // Truncate snippet text at sentence boundary for cleaner context
      const maxSnippetChars = 250;
      const text = truncateAtSentence(snippet.text, maxSnippetChars);

      selected.push({ ...snippet, text });
    }

    return selected;
  }

  /**
   * Format the recall block as markdown for systemPromptAddition.
   */
  private formatBlock(snippets: RecallSnippet[], durationMs: number): string {
    const lines: string[] = [
      "## 🧠 Recalled Context",
      `Automatically retrieved based on the user's message. Use for context; search for more detail if needed.`,
      "",
    ];

    for (const snippet of snippets) {
      lines.push(`- **[${snippet.reference}]** ${snippet.text}`);
    }

    // Timing footer — visible to the agent so it can monitor recall latency
    lines.push("");
    lines.push(`*Recall: ${snippets.length} snippets in ${durationMs}ms*`);

    return lines.join("\n");
  }
}

/**
 * Truncate text at a sentence boundary, falling back to word boundary.
 * Produces cleaner context than hard character cuts.
 */
function truncateAtSentence(text: string, maxChars: number): string {
  if (text.length <= maxChars) return text;

  // Try to find a sentence boundary within the budget
  const truncated = text.slice(0, maxChars);
  const lastSentenceEnd = Math.max(
    truncated.lastIndexOf(". "),
    truncated.lastIndexOf(".\n"),
    truncated.lastIndexOf("! "),
    truncated.lastIndexOf("? "),
  );

  // If we found a sentence boundary in the last 40% of the budget, use it
  if (lastSentenceEnd > maxChars * 0.6) {
    return truncated.slice(0, lastSentenceEnd + 1).trim();
  }

  // Fall back to word boundary
  const lastSpace = truncated.lastIndexOf(" ");
  if (lastSpace > maxChars * 0.7) {
    return truncated.slice(0, lastSpace).trim() + "…";
  }

  // Hard cut as last resort
  return truncated.trim() + "…";
}

/**
 * Convert a simple glob pattern to a RegExp.
 * Supports * (any chars except /) and ** (any chars including /).
 *
 * Handles edge case where double-star at start of pattern should match paths
 * that begin directly with the next segment (no leading /).
 * e.g., a glob like double-star/daily-notes/double-star should match both
 * "memory/daily-notes/2026-02/file.md" and "daily-notes/2026-02/file.md".
 */
function globToRegex(pattern: string): RegExp {
  const escaped = pattern
    .toLowerCase()
    .replace(/[.+^${}()|[\]\\]/g, "\\$&") // Escape regex special chars
    .replace(/\*\*/g, "§DOUBLESTAR§")      // Temp placeholder for **
    .replace(/\*/g, "[^/]*")               // * = anything except /
    .replace(/§DOUBLESTAR§/g, ".*");       // ** = anything including /

  // If pattern starts with .*/ (from **/) make the leading path optional
  // so it also matches paths that start directly with the segment.
  // e.g., ".*/daily-notes/.*" → "(.*/)?daily-notes/.*"
  const adjusted = escaped.startsWith(".*/")
    ? "(.*/)?".concat(escaped.slice(3))
    : escaped;

  return new RegExp(adjusted + "$", "i");
}

/**
 * Normalize a snippet reference for deduplication.
 * Strips collection prefixes, normalizes paths so the same file from
 * different sources (memory-search vs QMD vs LCM) merges properly.
 */
function normalizeDedupeKey(snippet: RecallSnippet): string {
  let key = snippet.reference.toLowerCase();

  // Strip leading path components that are collection/workspace prefixes
  // e.g., "memory/daily-notes/2026-02.md" should match
  //        "daily-notes/2026-02.md" from a different source
  // But keep lcm: prefixed refs distinct (they're not files)
  if (key.startsWith("lcm:")) return key;

  // Normalize common path prefixes — strip workspace/collection roots
  // Handles absolute paths like "/home/user/workspace/..." and relative "clawd/..." (QMD collection prefix)
  key = key
    .replace(/^(.*\/)?clawd\//, "")
    .replace(/^(.*\/)?workspace\//, "")
    .replace(/^(.*\/)?personal\//, "personal/");

  // Add line range to key if present to distinguish different snippets from same file
  if (snippet.lineStart) {
    key += `:${snippet.lineStart}`;
  }

  return key;
}

/**
 * Detect casual/phatic messages that won't benefit from recall.
 * Greetings, check-ins, acknowledgements — these produce noisy results.
 *
 * v0.3 fix: If the message contains substantive entities (proper nouns,
 * domain terms, abbreviations), it's NOT casual even if it starts with
 * a phatic pattern. "How's OathKeeper going?" is a real query.
 */
function isCasualMessage(text: string): boolean {
  const lower = text.toLowerCase().trim();

  // Too short to be meaningful — but only if no entities present
  if (lower.length < 15) {
    // Even very short messages might be entity lookups: "OathKeeper?" "PR status"
    return !hasSubstantiveEntities(text.trim());
  }

  // Common phatic patterns
  const casualPatterns = [
    /^(hey|hi|hello|yo|sup|morning|evening|night)\b/,
    /^(thanks|thank you|thx|ty)\b/,
    /^(ok|okay|sure|got it|sounds good|perfect|nice|great|cool|awesome)\b/,
    /^(yes|no|yep|nope|yeah|nah)\b/,
    /^how('s it)?\s*(goes|going|are you)/,
    /^just\s*(checking|poking|following)/,
    /^(you (there|around|working|awake|back))/,
    /^what('s)?\s*up\??$/,
  ];

  const matchesCasualPattern = casualPatterns.some((p) => p.test(lower));

  if (!matchesCasualPattern) return false;

  // Pattern matched — but check if there are substantive entities after the casual prefix.
  // "How's OathKeeper going?" matches the casual pattern but contains "OathKeeper".
  // In that case, it's NOT casual — the entity is the real query.
  return !hasSubstantiveEntities(text.trim());
}

/**
 * Detect heartbeat messages.
 */
function isHeartbeatMessage(text: string): boolean {
  const lower = text.toLowerCase().trim();
  return (
    lower.startsWith("read heartbeat.md") ||
    lower.includes("heartbeat_ok") ||
    lower.includes("follow it strictly. do not infer or repeat old tasks")
  );
}
