// @clawdactual/openclaw-recall/src/types.ts
/**
 * Core types for the Recall plugin.
 */

/**
 * A single recalled snippet — the atomic unit of recall output.
 */
export interface RecallSnippet {
  /** Which source produced this snippet */
  source: string;
  /** File path or identifier */
  path: string;
  /** The recalled text content */
  text: string;
  /** Relevance score (0-1) */
  score: number;
  /** Start line in source file (if applicable) */
  lineStart?: number;
  /** End line in source file (if applicable) */
  lineEnd?: number;
  /** Formatted source reference for display (e.g., "memory/2026-03-15.md:42-48") */
  reference: string;
}

/**
 * A recall source — searches one backend and returns snippets.
 */
export interface RecallSource {
  /** Source identifier */
  readonly name: string;
  /** Search this source and return scored snippets */
  search(query: string, minScore: number): Promise<RecallSnippet[]>;
}

/**
 * The assembled recall block, ready for injection.
 */
export interface RecallResult {
  /** Formatted markdown block for systemPromptAddition */
  block: string;
  /** Estimated token count of the block */
  estimatedTokens: number;
  /** How many snippets were included */
  snippetCount: number;
  /** How many total results were considered (before budget fitting) */
  totalResults: number;
  /** Which query extraction strategy was used */
  queryStrategy: string;
  /** Time taken in ms */
  durationMs: number;
}
